pulsar-client/src/main/java/org/apache/pulsar/client/impl/AutoClusterFailover.java

```diff
@@ -0,0 +1,234 @@
+                        primaryRecoverTimestamp = System.currentTimeMillis();
+                    } else if (System.currentTimeMillis() - primaryRecoverTimestamp >= switchBackDelayMs) {
+                        log.info("Current Pulsar service is secondary: {}, "
+                                        + "the primary service: {} has been recover for {} ms, "
+                                        + "switch back to the primary service",
+                                currentPulsarServiceUrl, primary, System.currentTimeMillis() - primaryRecoverTimestamp);
+                        try {
+                            pulsarClient.updateServiceUrl(primary);
+                            currentPulsarServiceUrl = primary;
+                        } catch (PulsarClientException e) {
+                            log.error("Current Pulsar service is secondary: {}, "
+                                    + "failed to switch back to primary service URL ", currentPulsarServiceUrl, e);
+                        }
+                    }
+                }
+            }
+        }, 30_000, 30_000);
+
+    }
+
+    @Override
+    public String getServiceUrl() {
+        return this.currentPulsarServiceUrl;
+    }
+
+    @Override
+    public void close() {
+        this.timer.cancel();
+    }
+
+    private boolean probeAvailable(String url, int timeout) {
+        try {
+            String hostAndPort = parseHostAndPort(url);
+            if (Strings.isNullOrEmpty(hostAndPort)) {
+                return false;
+            }
+
+            Socket socket = new Socket();
+            socket.connect(new InetSocketAddress(parseHost(hostAndPort), parsePort(hostAndPort)), timeout);
+            socket.close();
+            return true;
+        } catch (Exception e) {
+            return false;
+        }
+    }
+
+    private static String parseHostAndPort(String url) {
+        if (Strings.isNullOrEmpty(url) || !url.startsWith("pulsar")) {
+            throw new IllegalArgumentException("'" + url + "' isn't an Pulsar service URL");
+        }
+
+        int uriSeparatorPos = url.indexOf("://");
+        if (uriSeparatorPos == -1) {
+            throw new IllegalArgumentException("'" + url + "' isn't an URI.");
+        }
+        return url.substring(uriSeparatorPos + 3);
+    }
+
+    private static String parseHost(String hostAndPort) {
+        int portSeparatorPos = hostAndPort.indexOf(":");
+        if (portSeparatorPos == -1) {
+            throw new IllegalArgumentException("'" + hostAndPort + "' isn't an URI.");
+        }
+        return hostAndPort.substring(0, portSeparatorPos);
+    }
+
+    private static Integer parsePort(String hostAndPort) {
+        int portSeparatorPos = hostAndPort.indexOf(":");
+        if (portSeparatorPos == -1) {
+            throw new IllegalArgumentException("'" + hostAndPort + "' isn't an URI.");
+        }
+        return Integer.valueOf(hostAndPort.substring(portSeparatorPos+1));
+    }
```

Comment:
Would it be better to use `URL` class to parse a URL and you can get the host and port easily.

Reply:
I think `URI` could be better here, because `URL` assumes a "known" protocol scheme.

Reply:
@hangc0276 please address this comment

Reply:
done